### DISCUSSION TITLE
Use functions instead of classes to define JavaScript components

### DISCUSSION THREAD

**Discussion by luchian** [twbs/bootstrap#34575] [General]
This requires a major overhaul but would greatly simplify component definitions. Not having to worry about the context of `this` would ease understanding of how components work and bolster confidence when adding features or items. There are performance-related worries but it looks like those do not translate to real world use cases. React hooks are created using functions and have been adopted without major issues.

As I said, this would be a major undertaking so I was not able to make a pull request but a functional component would resemble existing ones minus the `this` keyword.
``` javascript
function Alert(element) {
  ...

  function close(element) {
    const rootElement = element ? _getRootElement(element) : _element
    const customEvent = _triggerCloseEvent(rootElement)

    if (customEvent === null || customEvent.defaultPrevented) {
      return
    }

    _removeElement(rootElement)
  }

  ...
}
```
Here are some articles that explore this subject in greater detail:
https://css-tricks.com/the-flavors-of-object-oriented-programming-in-javascript/
https://www.toptal.com/javascript/es6-class-chaos-keeps-js-developer-up

**Comment by jjoel1630** [comment]
Hello @luchian. This would be my first time contributing to an open source project, so do you think that I could be assigned to this issue? I have sufficient knowledge in JavaScript.

**Comment by luchian** [comment]
Hi Joel! It's great you're starting to contribute to open source projects. I'm not a member of the Bootstrap organization so I cannot assign issues, but you could try opening a pull request once you have some code ready. A discussion about your commits will take place and if everything is in order it will be merged into the main branch.

**Comment by jjoel1630** [comment]
Thanks! So I should start working on the code and then open a pull request, is what you are saying?

**Comment by alpadev** [comment]
IDK in the opening post you mention React hooks and how well adopted they are and functional components which are two different things actually.. For how long are hooks a thing yet? 2-3 years? I'm pretty sure in the majority of projects they are rarely used yet, if even. And they supplemented functional components, so functional components became actually useful. AFAIK React is still supporting (and has since the beginning) to extend classes to define your components, and not deprecated this pattern yet.

However, Bootstrap isn't React to begin with, and I don't think that this will ever being a goal for us to achieve. The purpose of our JS is to add necessary functionality to our components, not to be a fully fledged JS framework. There are always people arguing about OOP vs FP and from what I can tell both camps have their reasons to favor one over the other and thats perfectly fine. 

The code base is using classes and people who contributed here lately seem to be ok with that approach. Throwing everything out just to rebuild things in a functional way - not sure if this is on our roadmap or anything we aim for. Also I don't see really the benefits here TBH. `this` used to be tricky at times, but with classes, bind and arrow functions this is like a thing from the past IMO.

I'm just one person responding here. So don't see this as a consensus on our team.

**Comment by jjoel1630** [comment]
@alpadev Thank you for your response. I completely understand what you are saying. It will be a pain to rebuild everything using functional programming. I just saw this issue and was willing to help, but I completely understand where you are going. If you are ever going to transition to functional programming just let me know. Thank you again.

**Comment by alpadev** [comment]
@jjoel1630 Glad there are people trying to help and I'm not saying this isn't something that is going to happen eventually in the future.

v5 of Bootstrap borrows a lot of old code from v4 and beyond still. We're in the process of refactoring things but from what I can tell, everyone of us doing this on a voluntary basis, having daily jobs that are not focused around Bootstrap. That makes it hard to really spend the amount of time to pull off a paradigm shift like this.

Another argument would be the need to backport things to v4, as we still trying to support that code base as well. To cripple everything means we need to implement things twice, which requires even more time and effort to achieve that.

Actually I was hoping for a response of someone who's on this project for a longer time, but I didn't want you to spend a lot of time already by trying to move into this direction.

Every one who is willing to contribute to improve things is greatly appreciated, but as it stands right now, I don't see that change to happen in a closer future.

**Comment by jjoel1630** [comment]
Yes, I completely understand. Whenever you are willing to make the change, just let me know. I will be happy to help you guys out. In the meantime, I will try to contribute by fixing other issues.

**Comment by luchian** [comment]
@alpadev I mentioned React only to provide a real-world use case of components defined using functions. Of course Bootstrap is different from React but this does not mean that practices that work well for one should not be implemented by the other. Classes vs. functions is a contentious matter yet one worth discussing as it may help establish a code base that is easier to maintain, update, and scale.

From a technical standpoint, there are ways of ensuring `this` behaves predictably, but the fact remains that using `bind` or arrow functions is a convoluted way of doing what can easily be achieved without any drawbacks with functions. Simplicity in component definitions is important.

I understand that such a change is radical so I'll try to open a pull request once I have some code ready.

**Comment by luchian** [comment]
@jjoel1630 You can contribute by helping with any of the open issues or by submitting for review your own code implementing some feature.
